### PR TITLE
Develop

### PR DESCRIPTION
**Pull request by surajsbharadwaj**
### Description

Please include a summary of the change and which issue is fixed. Also, please include the motivation and context for the change. List any dependencies that are required for this change.

**Check the relevant boxes:**
- [x] Bug fix (nonbreaking change that fixes an issue)
- [x] New feature (nonbreaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Examples or tests (addition or updates of examples or tests)
- [x] Documentation update
- [x] CI related update (pipeline, etc.)

### Checklist

- [ ] If relevant, a test for the change has been added or updated as part of this PR.
- [ ] If relevant, documentation for the change has been added or updated as part of this PR.

### Merge

- Merge using "Squash and merge".
- Make sure to use a relevant [conventional commit](https://www.conventionalcommits.org/) message that is based on the PR contents. The commit message determines whether a new version of the modules needs to be released, and if so, which semver number to use).

### CODE REVIEW

**Comment by Beilmann** [comment]
I do not think that you want to replace .pre-commit-config.yaml, right?

**Comment by vburckhardt** [comment]
@SirSpidey - Hi Allen, do you want to take a look at the documentation before this gets merged?

**Review by Beilmann — APPROVED** [review]

**Comment by SirSpidey** [comment]
@surajsbharadwaj @vburckhardt In addition to content suggestions in this PR, I updated the descriptions in `variables.tf` in https://github.com/terraform-ibm-modules/terraform-ibm-powervs-infrastructure/pull/35 to merge into this PR.

After those variable descriptions are approved, I could look at the other variable descriptions.

**Comment by surajsbharadwaj** [comment]
> @surajsbharadwaj @vburckhardt In addition to content suggestions in this PR, I updated the descriptions in `variables.tf` in #35 to merge into this PR.
> 
> After those variable descriptions are approved, I could look at the other variable descriptions.

Thank you, i included them in the commit. Thank you

File: submodules/power_cloudconnection_attach/variables.tf
Comment (vburckhardt) [inline submodules/power_cloudconnection_attach/variables.tf:23]: type should be "number"

File: submodules/power_cloudconnection_create/README.md
Comment (vburckhardt) [inline submodules/power_cloudconnection_create/README.md:8]: Indentation lost in commit

File: submodules/power_cloudconnection_create/README.md
Comment (vburckhardt) [inline submodules/power_cloudconnection_create/README.md:3]: It would be worth linking out to relevant Cloud documentation on Cloud Connections

File: variables.tf
Comment (vburckhardt) [inline variables.tf:99]: There should be a default here set to null - otherwise this forces users of the module to have a jump box even if they opt out of deployment the power managemenr services.

File: variables.tf
Comment (vburckhardt) [inline variables.tf:119]: Perhaps document where those dns server come from (eg: IBM Cloud doc for the 161 page) and rational for adding quad9 dns here (have we checked this is a preferred DNS provider as per corp policies?)

File: variables.tf
Comment (vburckhardt) [inline variables.tf:100]: This could be the IP of any jump box reachable from the machine running the terraform logic. Could be a private IP in the scenario where this code is running from a vsi within the VPC) - although there is probably no need for a jump box in that scenario in theory (but code assumes there is one in practice)

File: submodules/power_management_services_setup/main.tf
Comment (vburckhardt) [inline submodules/power_management_services_setup/main.tf:11]: Don't think it's a good idea to default to root here. Could we make this configurable?

File: submodules/power_management_services_setup/main.tf
Comment (vburckhardt) [inline submodules/power_management_services_setup/main.tf:12]: It'd be desirable to add bastion_user and bastion_private_key in the block, so that users can configure different keys for bastion and remove host, and different user. This would come handy to further segregate access.

File: submodules/power_management_services_setup/main.tf
Comment (vburckhardt) [inline submodules/power_management_services_setup/main.tf:23]: Could you extract those scripts out as file, and use the script block to reference them? There are static checks on the script file in the pipeline that can help catch issues.

File: examples/basic/variables.tf
Comment (vburckhardt) [inline examples/basic/variables.tf:97]: Remove all inputs that are not designed to be changed. Make them locals instead if needed.

File: submodules/power_management_services_setup/README.md
Comment (vburckhardt) [inline submodules/power_management_services_setup/README.md:1]: This module seems to have very specific expectations in term of flavor of vsi for jump box and services (eg: suse), but 
1. This is not documented explicitly anywhere
2. There is no validation block (remote-exec) that the hosts are aligned with requirement (eg: could do a uname -a, etc to validate and return a nice error message)

File: submodules/initial_validation/README.md
Comment (vburckhardt) [inline submodules/initial_validation/README.md:1]: I think this logic would fit better within the root module - see example of the pattern at https://github.com/terraform-ibm-modules/terraform-ibm-resource-group/blob/a48ba6581c714d31aca070d330bd9310141ad977/main.tf#L8-L15

File: examples/standard/variables.tf
Comment (vburckhardt) [inline examples/standard/variables.tf:110]: Could you add a validation block to check inputs

File: examples/standard/variables.tf
Comment (vburckhardt) [inline examples/standard/variables.tf:111]: number instead of string?

**Review by vburckhardt — CHANGES_REQUESTED** [review]
Thanks for the PR. Some comments through a first pass. We may need further reviews as it's a very large PR (small PRs are better)

File: submodules/power_management_services_setup/README.md
Comment (vburckhardt) [reply submodules/power_management_services_setup/README.md:1]: It'd also be good to have one of the example using the service_setup, currently there is no example, nor tests against that logic.

File: submodules/initial_validation/README.md
Comment (Beilmann) [reply submodules/initial_validation/README.md:1]: main.tf is not the right place. The workflow should fail in terraform plan, if input variables do not fit to each other. Decided to make a separate module and include all the logic to verify input + environments BEFORE something will be created.

File: submodules/power_management_services_setup/README.md
Comment (Beilmann) [reply submodules/power_management_services_setup/README.md:1]: - standard and imm-catalog/standard-solution examples utilise management submodule.
- the only use case where management submodule is not used is the basic example (due to limitations in landscape prerequisites, like missing transit gateway, and VSIs on VPC)
- setup management services --> proxy is more or less mandatory, since otherwise you do not have access to public internet from PowerVS.

File: submodules/power_management_services_setup/main.tf
Comment (Beilmann) [reply submodules/power_management_services_setup/main.tf:11]: in the image, other users (we could potentially use) are not available after the installation. And btw. all SAP installations from bottom to top are executed as root.

File: variables.tf
Comment (Beilmann) [reply variables.tf:100]: It could be any IP that is reachable from TF engine and allows access to the environment. Practically, this is public IP.

File: variables.tf
Comment (Beilmann) [reply variables.tf:99]: the only scenario where access host is not needed is basic. Otherwise, creating landscape that you cannot reach, does not make sense in my oppinion

File: submodules/power_cloudconnection_create/README.md
Comment (Beilmann) [reply submodules/power_cloudconnection_create/README.md:3]: there would be general documentation on cloud doc that describes this automation (and how to use this manually)

File: submodules/power_management_services_setup/main.tf
Comment (Beilmann) [reply submodules/power_management_services_setup/main.tf:12]: in general, good to segregate, but:
- if we would use another user, we must create it from root first
- the whole SAP installation is executed as root

File: submodules/power_management_services_setup/main.tf
Comment (Beilmann) [reply submodules/power_management_services_setup/main.tf:23]: would be good. Can you give an example?

File: examples/basic/README.md
Comment (surajsbharadwaj) [reply examples/basic/README.md:6]: because the point 2 private networks is a subpoint for Creates a PowerVS. And terraform docs didnt support the sub bullet

File: submodules/initial_validation/README.md
Comment (vburckhardt) [reply submodules/initial_validation/README.md:1]: The logic above would cause to fail at terraform plan

File: submodules/power_management_services_setup/README.md
Comment (vburckhardt) [reply submodules/power_management_services_setup/README.md:1]: @Beilmann - It'd be good to have this in one example that is run through ci pipeline so that regressions can be caught early in this area. I do not think ibm-catalog/standard-solution is currently run in ci pipeline (not referenced from https://github.com/terraform-ibm-modules/terraform-ibm-powervs-infrastructure/blob/develop/tests/pr_test.go) so either we add this to the pr_test or create an additional example leveraging the slz module. The slz module is already available at https://github.com/terraform-ibm-modules/terraform-ibm-landing-zone

Comment (vburckhardt) [reply submodules/power_management_services_setup/README.md:1]: We could have this example / test in a subsequent PR, I do not think this block this one. We should just have some more documentation and checks on expectations in this PR tough.

File: variables.tf
Comment (surajsbharadwaj) [reply variables.tf:106]: done

File: README.md
Comment (surajsbharadwaj) [reply README.md:16]: making changes

File: README.md
Comment (surajsbharadwaj) [reply README.md:1]: will remove it

File: examples/basic/variables.tf
Comment (surajsbharadwaj) [reply examples/basic/variables.tf:97]: need to think about this, as you want to execute the whole module, including os initialization

File: submodules/power_management_services_setup/README.md
Comment (surajsbharadwaj) [reply submodules/power_management_services_setup/README.md:1]: Hmm right.

File: submodules/initial_validation/README.md
Comment (surajsbharadwaj) [reply submodules/initial_validation/README.md:1]: Right, But it is good to keep it in a separate module which tracks all variable validation right ?

File: examples/standard/variables.tf
Comment (surajsbharadwaj) [reply examples/standard/variables.tf:110]: Sure will include it

File: examples/standard/variables.tf
Comment (surajsbharadwaj) [reply examples/standard/variables.tf:111]: Done

File: submodules/power_cloudconnection_attach/variables.tf
Comment (surajsbharadwaj) [reply submodules/power_cloudconnection_attach/variables.tf:23]: DOne converted

File: submodules/power_cloudconnection_create/README.md
Comment (surajsbharadwaj) [reply submodules/power_cloudconnection_create/README.md:8]: Will resolve it

File: variables.tf
Comment (surajsbharadwaj) [reply variables.tf:99]: Added

File: submodules/power_management_services_setup/main.tf
Comment (surajsbharadwaj) [reply submodules/power_management_services_setup/main.tf:11]: Need root.

File: submodules/power_management_services_setup/main.tf
Comment (surajsbharadwaj) [reply submodules/power_management_services_setup/main.tf:23]: Yes its good to put it in a script will work on it and push it

File: submodules/initial_validation/README.md
Comment (vburckhardt) [reply submodules/initial_validation/README.md:1]: @surajsbharadwaj - It's the same. If you want to stick with a different module for validation - could you just remove the empty tf files (main, output)

File: submodules/initial_validation/versions.tf
Comment (vburckhardt) [inline submodules/initial_validation/versions.tf:2]: subModule -> sub-module or submodule

File: examples/basic/README.md
Comment (vburckhardt) [reply examples/basic/README.md:6]: I don't believe terraform docs would be touching this part of the README. It only touches the pieces between  blocks

File: examples/basic/variables.tf
Comment (vburckhardt) [reply examples/basic/variables.tf:97]: Set it a locale in this scenario - the variable description currently indicate to not touch its value "Do not configure proxy in this example by default."

**Review by surajsbharadwaj — COMMENTED** [review]
Please merge to main

File: README.md
Comment (vburckhardt) [reply README.md:14]: @surajsbharadwaj - Still an issue - do you have remaining local changes not pushed?

File: examples/basic/README.md
Comment (vburckhardt) [reply examples/basic/README.md:6]: Still an issue

File: examples/basic/variables.tf
Comment (vburckhardt) [reply examples/basic/variables.tf:97]: Still an issue

File: examples/ibm-catalog/standard-solution/variables.tf
Comment (vburckhardt) [inline examples/ibm-catalog/standard-solution/variables.tf:2]: The public name is "Secure infrastructure on VPC for regulated industries" instead of "Secure Landing Zone"

File: examples/ibm-catalog/standard-solution/variables.tf
Comment (vburckhardt) [inline examples/ibm-catalog/standard-solution/variables.tf:9]: Could we add a validation block here to check the values

File: examples/ibm-catalog/standard-solution/variables.tf
Comment (vburckhardt) [inline examples/ibm-catalog/standard-solution/variables.tf:88]: Tag -> tag

File: examples/standard/variables.tf
Comment (vburckhardt) [reply examples/standard/variables.tf:110]: Still not addressed

File: submodules/power_cloudconnection_create/README.md
Comment (vburckhardt) [reply submodules/power_cloudconnection_create/README.md:3]: I still think we should make it easier for people consuming directly from github.com

File: submodules/power_management_services_setup/README.md
Comment (vburckhardt) [reply submodules/power_management_services_setup/README.md:1]: @surajsbharadwaj - could the action be tracked in an issue ticket at https://github.com/terraform-ibm-modules/terraform-ibm-issue-tracker so that it does not fall through the cracks?

File: submodules/power_management_services_setup/README.md
Comment (vburckhardt) [inline submodules/power_management_services_setup/README.md:6]: Indentation issue

File: submodules/power_management_services_setup/main.tf
Comment (vburckhardt) [reply submodules/power_management_services_setup/main.tf:17]: Ok let's review in future when the case emerges

File: submodules/power_management_services_setup/scripts/squid_proxy.sh
Comment (vburckhardt) [inline submodules/power_management_services_setup/scripts/squid_proxy.sh:95]: indentation

File: submodules/power_management_services_setup/scripts/squid_proxy.sh
Comment (vburckhardt) [inline submodules/power_management_services_setup/scripts/squid_proxy.sh:2]: add `set -euo pipefail` . See https://gist.github.com/mohanpedala/1e2ff5661761d3abd0385e8223e16425

File: submodules/power_management_services_setup/scripts/squid_proxy.sh
Comment (vburckhardt) [inline submodules/power_management_services_setup/scripts/squid_proxy.sh:106]: could we add a max number of executions here on those loops + sleep between each loop iteration?

File: submodules/power_management_services_setup/variables.tf
Comment (vburckhardt) [inline submodules/power_management_services_setup/variables.tf:12]: SSh -> ssh

Comment (vburckhardt) [reply submodules/power_management_services_setup/variables.tf:12]: "It will not be uploaded" - is this fully accurate - is it uploaded to the bastion host at minimum however?

File: variables.tf
Comment (vburckhardt) [reply variables.tf:139]: Still not addressed

**Review by vburckhardt — CHANGES_REQUESTED** [review]
Thanks for the change. Here are the comments for the 2nd pass

File: examples/ibm-catalog/standard-solution/variables.tf
Comment (surajsbharadwaj) [reply examples/ibm-catalog/standard-solution/variables.tf]: Fixed

File: submodules/power_management_services_setup/scripts/squid_proxy.sh
Comment (Beilmann) [reply submodules/power_management_services_setup/scripts/squid_proxy.sh:106]: @surajsbharadwaj: let's do it in next iteration.
@vburckhardt: is that ok to do it in next iteration?

File: examples/ibm-catalog/standard-solution/variables.tf
Comment (Beilmann) [reply examples/ibm-catalog/standard-solution/variables.tf:9]: @surajsbharadwaj: let's do it in next iteration
@vburckhardt: is that ok to do it in next iteration?

File: submodules/power_management_services_setup/variables.tf
Comment (Beilmann) [reply submodules/power_management_services_setup/variables.tf:12]: @vburckhardt: ssh key is not uploaded. we use ssh agent mechanism

Comment (surajsbharadwaj) [reply submodules/power_management_services_setup/variables.tf:12]: Yes accurate. The key is not stored anywhere

File: README.md
Comment (surajsbharadwaj) [reply README.md:14]: Fixed. As it was manual work

File: submodules/power_management_services_setup/README.md
Comment (surajsbharadwaj) [reply submodules/power_management_services_setup/README.md:1]: @Beilmann can you comment here ?

File: examples/standard/variables.tf
Comment (surajsbharadwaj) [reply examples/standard/variables.tf:110]: I didnt handle it because the apply itself throws an error saying not  right values, But now i realized that, it can be caught during plan itself

File: submodules/initial_validation/versions.tf
Comment (surajsbharadwaj) [reply submodules/initial_validation/versions.tf]: Fixed

File: submodules/power_management_services_setup/scripts/squid_proxy.sh
Comment (surajsbharadwaj) [reply submodules/power_management_services_setup/scripts/squid_proxy.sh:106]: next PR please

Comment (vburckhardt) [reply submodules/power_management_services_setup/scripts/squid_proxy.sh:106]: Ok - could you track this in an issue at https://github.com/terraform-ibm-modules/terraform-ibm-issue-tracker so that it does not get lost

File: examples/ibm-catalog/standard-solution/variables.tf
Comment (vburckhardt) [reply examples/ibm-catalog/standard-solution/variables.tf:9]: Yes - just track the action at https://github.com/terraform-ibm-modules/terraform-ibm-issue-tracker

File: examples/basic/variables.tf
Comment (surajsbharadwaj) [reply examples/basic/variables.tf]: DOne

File: examples/standard/variables.tf
Comment (surajsbharadwaj) [reply examples/standard/variables.tf:110]: NOt required, In plan the provider itself throws an error if not valid

File: submodules/power_management_services_setup/README.md
Comment (Beilmann) [reply submodules/power_management_services_setup/README.md:1]: @surajsbharadwaj - track the issue for extending basic example to deploy SLZ available in public git and put powerVS infrastructure on top of it. Then we can use cloud connections and also deploy management services.

**Comment by vburckhardt** [comment]
Ok folks, let's go with this first cut and continue in a subsequent PR. Tracked the remaining items at https://github.com/terraform-ibm-modules/terraform-ibm-issue-tracker/issues/45

**Review by vburckhardt — APPROVED** [review]

**Review by vburckhardt — APPROVED** [review]

**Pull request merged**
